janus_server/src/aggregator/aggregate_share.rs

```diff
@@ -357,71 +358,71 @@ where
+    // Each such row consumes one unit of batch lifetime (§4.6).
+    let intersecting_intervals: Vec<_> = match task.role {
+        Role::Leader => tx
+            .find_collect_jobs_jobs_intersecting_interval::<L, A>(task.id, collect_interval)
+            .await?
+            .into_iter()
+            .map(|job| job.batch_interval)
+            .collect(),
+
+        Role::Helper => tx
+            .find_aggregate_share_jobs_intersecting_interval::<L, A>(task.id, collect_interval)
+            .await?
+            .into_iter()
+            .map(|job| job.batch_interval)
+            .collect(),
+
+        _ => panic!("Unexpected task role {:?}", task.role),
+    };
+
+    // Check that all intersecting collect intervals are equal to this collect interval.
+    if intersecting_intervals
         .iter()
-        .map(|v| {
-            Interval::new(v.unit_interval_start, task.min_batch_duration)
-                .map_err(|e| datastore::Error::User(e.into()))
-        })
-        .collect::<Result<_, datastore::Error>>()?;
-
-    let overlaps = tx
-        .get_aggregate_share_job_counts_for_intervals(task.id, task.role, &intervals)
-        .await?;
-
-    for (unit_interval, consumed_batch_lifetime) in overlaps {
-        if consumed_batch_lifetime == task.max_batch_lifetime {
-            debug!(
-                ?task.id, ?unit_interval,
-                "refusing aggregate share request because lifetime for batch unit has been consumed"
-            );
-            return Err(datastore::Error::User(
-                Error::BatchLifetimeExceeded(task.id).into(),
-            ));
-        }
-        if consumed_batch_lifetime > task.max_batch_lifetime {
-            error!(
-                ?task.id, ?unit_interval,
-                "batch unit lifetime has been consumed more times than task allows"
-            );
-            panic!("batch unit lifetime has already been consumed more times than task allows");
-        }
+        .any(|interval| interval != &collect_interval)
+    {
+        return Err(datastore::Error::User(
+            Error::BatchInvalid(collect_interval, task.id).into(),
+        ));
     }
 
+    // Check that the batch lifetime is being consumed appropriately.
+    let max_batch_lifetime: usize = task.max_batch_lifetime.try_into()?;
+    if intersecting_intervals.len() == max_batch_lifetime {
+        debug!(
+            task_id = ?task.id, ?collect_interval,
+            "Refusing aggregate share request because batch lifetime has been consumed"
+        );
+        return Err(datastore::Error::User(
+            Error::BatchLifetimeExceeded(task.id).into(),
+        ));
+    }
+    if intersecting_intervals.len() > max_batch_lifetime {
+        error!(
+            task_id = ?task.id, ?collect_interval,
+            "Batch lifetime has been consumed more times than task allows"
+        );
+        panic!("Batch lifetime has been consumed more times than task allows");
```

Comment:
Hmm... ultimately, the first highlighted condition rolls up into the generic request-level metric tracking (since all uses of this function are ultimately from some HTTP endpoint handler). The second panics because it should be "impossible" (up to someone modifying the task in the datastore, anyway) -- we probably should add a metric there, and maybe stop panicking.

The new condition around no-partial-overlaps also rolls up into the generic request-level metric tracking, too.

But -- I just noticed the generic request-level tracking doesn't track the reason for failure (e.g. it doesn't record an attribute that is a specific error code, just an "status" of `error`). I think I'm going to expand this in a followup PR, and make the changes to the currently-panicking condition as well.